azkaban-exec-server/src/main/java/azkaban/execapp/FlowRunnerManager.java

```diff
@@ -1053,4 +1056,63 @@ public void shutdown() {
       this.scheduler.shutdownNow();
     }
   }
+
+  private class PollingCriteria {
+
+    private final Props azkabanProps;
+    private final SystemMemoryInfo memInfo = SERVICE_PROVIDER.getInstance(SystemMemoryInfo.class);
+    private final OsCpuUtil cpuUtil = SERVICE_PROVIDER.getInstance(OsCpuUtil.class);
```

Comment:
Why do you use provider instead of normal injection?

Reply:
This is a non-static inner class of FlowRunnerManager. The injector cannot (as far as I know) inject this type of classes, it would have to be a static inner class but it uses FlowRunnerManager’s logger, getMaxNumRunningFlows() and getNumRunningFlows() which are non-static members. PollingCriteria behaves similarly to PollingService class, I just followed the latter’s pattern.